controlplane/kubeadm/internal/cluster_test.go

```diff
@@ -178,6 +154,178 @@ func TestGetMachinesForCluster(t *testing.T) {
 	g.Expect(machines).To(HaveLen(1))
 }
 
+func TestGetWorkloadCluster(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	// create an envtest for all subtests
+	testEnv := helpers.NewTestEnvironment()
+	defer func() {
+		if testEnv != nil {
+			g.Expect(testEnv.Stop()).To(Succeed(), "stopping envtest")
+		}
+	}()
+	go func() {
+		g.Expect(testEnv.StartManager()).To(Succeed())
+	}()
+
+	// Create an etcd secret with valid certs
+	key, err := certs.NewPrivateKey()
+	g.Expect(err).ToNot(HaveOccurred())
+	cert, err := getTestCACert(key)
+	g.Expect(err).ToNot(HaveOccurred())
+	etcdSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-etcd",
+			Namespace: "my-namespace",
+		},
+		Data: map[string][]byte{
+			secret.TLSCrtDataName: certs.EncodeCertPEM(cert),
+			secret.TLSKeyDataName: certs.EncodePrivateKeyPEM(key),
+		},
+	}
+
+	// Create kubeconfig secret
+	// Store the envtest config as the contents of the kubeconfig secret.
+	// This way we are using the envtest environment as both the
+	// management and the workload cluster.
+	testEnvKubeconfig := kubeconfig.FromEnvTestConfig(testEnv.GetConfig(), &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster",
+			Namespace: "my-namespace",
+		},
+	})
+	kubeconfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-kubeconfig",
+			Namespace: "my-namespace",
+		},
+		Data: map[string][]byte{
+			secret.KubeconfigDataName: testEnvKubeconfig,
+		},
+	}
+	clusterKey := client.ObjectKey{
+		Namespace: "my-namespace",
+		Name:      "my-cluster",
+	}
+
+	tests := []struct {
+		name       string
+		clusterKey client.ObjectKey
+		objs       []runtime.Object
+		expectErr  bool
+	}{
+		{
+			name:       "returns a workload cluster",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{etcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
```

Comment:
Passing in DeepCopy objs since below when we do `testEnv.Create` it seems to update the object in place and as part of that it sets the `resourceVersion` so in the next table test if we use the same object, `testEnv.Create` will throw an error: _resourceVersion should not be set on objects to be created_

Reply:
yes, this is one of the differences between fakeclient and a real client.